pkg/recommendation/minmaxcapacity.go

```diff
@@ -0,0 +1,203 @@
+// If there is any error during validation then it returns an error.
+func (r *MinMaxCapacityRequest) BuildAndVerifyOrError() (
+	*MinMaxCapacityRequest, error) {
+	if r.RaidGroupConfig == nil {
+		r.RaidGroupConfig = types.GetDefaultRaidGroupConfig()
+	}
+	err := r.RaidGroupConfig.Validate()
+	if err != nil {
+		return nil, err
+	}
+	return r, nil
+}
+
+// Get calculate min max and return it as a response.
+func (r *MinMaxCapacityRequest) Get() MinMaxCapacity {
+	deviceTypeNodeBlockDeviceMap := buildNodeBlockDeviceMap(*r.BlockDeviceList)
+	resultMap := make(map[string]*MinMaxCapacity)
+
+	deviceKind := []string{DeviceKindSSD, DeviceKindHDD}
+
+	// Get min max capacity for each type of block devices.
+	for _, kind := range deviceKind {
+
+		// Check nodeBlockDeviceListMap present for a device type not
+		// if not then skip fetching min max for that kind.
+		nodeBlockDeviceListMap, ok := deviceTypeNodeBlockDeviceMap[kind]
+		if !ok {
+			continue
+		}
+
+		// nodeCapacityDeviceCountMap contains block device capacity and count
+		// for that capacity for each node (block device is present in that node)
+		// After populating nodeBlockDeviceListMap with data this map with looks
+		// like this -
+		// Node-1
+		//     - [100GB, 4]
+		//     - [500GB, 6]
+		//     - [1TB, 4]
+		// Node-2
+		//     - [100GB, 1]
+		//     - [500GB, 8]
+		//     - [1TB, 2]
+		// Node-3
+		//     - [500GB, 5]
+		//     - [1TB, 5]
+		// Node-4
+		//     - [100GB, 8]
+		//     - [500GB, 3]
+		nodeCapacityDeviceCountMap := make(map[string]map[int64]int64)
+
+		// Iterating over nodeBlockDeviceListMap map and populating
+		// nodeCapacityDeviceCountMap
+		for nodeName, blockDeviceList := range nodeBlockDeviceListMap {
+
+			// If capacity count map not found for a node then create it.
+			capacityCountMap, capacityCountMapFound := nodeCapacityDeviceCountMap[nodeName]
+			if !capacityCountMapFound {
+				capacityCountMap = make(map[int64]int64)
+			}
+
+			// Iterate over blockDeviceList and populate
+			for _, blockDevice := range blockDeviceList {
+
+				// find capacity for each block device nodeBlockDeviceListMap
+				capacity, capacityFound := blockDevice.Capacity.AsInt64()
+				if !capacityFound {
+					// TODO handle this case
+					continue
+				}
+
+				// In capacity device count map if entry not found for any capacity then
+				// make it 0. If found then increase by 1.
+				_, ok := capacityCountMap[capacity]
+				if !ok {
+					capacityCountMap[capacity] = 0
+				}
+				capacityCountMap[capacity]++
+			}
+
+			// update nodeCapacityDeviceCountMap for each node.
+			nodeCapacityDeviceCountMap[nodeName] = capacityCountMap
+		}
+
+		result := &MinMaxCapacity{
+			MinCapacity: resource.Quantity{},
+			MaxCapacity: resource.Quantity{},
+		}
+
+		// Iterate over nodeCapacityDeviceCountMap and get min and max.
+		for _, capacityDeviceCountMap := range nodeCapacityDeviceCountMap {
+			for capacity, deviceCount := range capacityDeviceCountMap {
```

Comment:
It seems to me that a new capacity recommentation instance should be created per item i.e. on each iteration.
We can add these recommendations to a list as shown below. Then the final min & max can be returned.
```go
type CapacityRecommendationList []CapacityRecommendation
```
or
```go
type CapacityRecommendationMap map[string]CapacityRecommendation
```

Reply:
This is an intermediate map to calculate min-max capacity.